### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by J0hnCena**
A lot of general documentation was added. I did not fully understand how WidgetFlowRenderable worked, so any comments or tips for how to improve that section or even the overall pull request are welcome.

### CODE REVIEW

File: src/main/java/org/terasology/inGameHelp/InGameHelpCategoryRegistry.java [hunk 1]
```diff
@@ -17,7 +17,20 @@
 
 import org.terasology.inGameHelp.systems.HelpCategory;
 
+/**
+ * Interface for managing the different help categories
+ */
 public interface InGameHelpCategoryRegistry {
+
+	/**
+	 * @return a list of the help categories
```
Comment (skaldarnar) [inline src/main/java/org/terasology/inGameHelp/InGameHelpCategoryRegistry.java]: for the sake of completeness, please add one sentence in addition to `@return`, like 
```
"Retrieve the {@link org.terasology.inGameHelp.systems.HelpCategory}s previously registered to this registry."
```

File: src/main/java/org/terasology/inGameHelp/InGameHelpCategoryRegistry.java [hunk 2]
```diff
@@ -17,7 +17,20 @@
...
+	 */
     Iterable<HelpCategory> getCategories();
+
+	/**
+	 * Add a HelpCategory to the list
```
Comment (skaldarnar) [inline src/main/java/org/terasology/inGameHelp/InGameHelpCategoryRegistry.java]: ~to the list~ to this registry

File: src/main/java/org/terasology/inGameHelp/InGameHelpCategoryRegistry.java [hunk 3]
```diff
@@ -17,7 +17,20 @@
...
+	 *
+	 * @param category the HelpCategory to add
```
Comment (skaldarnar) [inline src/main/java/org/terasology/inGameHelp/InGameHelpCategoryRegistry.java]: either use "help category" in plain text or `{@link ...}` to the class.

File: src/main/java/org/terasology/inGameHelp/ItemsCategoryInGameHelpRegistry.java [hunk 4]
```diff
@@ -18,10 +18,24 @@
...
 
+/**
+ * Interface for dealing with the prefabs for the items
+ */
 public interface ItemsCategoryInGameHelpRegistry {
+
+    /**
+     * Add a Prefab and a variable amount of HelpItems
+     */
     void addKnownPrefab(Prefab prefab, HelpItem... helpItems);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/inGameHelp/ItemsCategoryInGameHelpRegistry.java]: this associates HelpItems with prefabs, i.e., one prefab can be connected with many help items.

File: src/main/java/org/terasology/inGameHelp/ItemsCategoryInGameHelpRegistry.java [hunk 5]
```diff
@@ -18,10 +18,24 @@
...
 
+    /**
+     * @return an Iterable containing the Prefabs
+     */
     Iterable<Prefab> getKnownPrefabs();
 
+    /**
+     * @param prefab the Prefab with which to get the HelpItems from
```
Comment (skaldarnar) [inline src/main/java/org/terasology/inGameHelp/ItemsCategoryInGameHelpRegistry.java]: imho does not get the help items from but _for_ the prefab

File: src/main/java/org/terasology/inGameHelp/components/HelpItem.java [hunk 6]
```diff
@@ -20,13 +20,31 @@
 import org.terasology.rendering.nui.widgets.browser.data.basic.HTMLLikeParser;
 import org.terasology.rendering.nui.widgets.browser.data.html.HTMLDocument;
 
+/**
+ * Interface for adding help information to a document
+ */
 public interface HelpItem {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/inGameHelp/components/HelpItem.java]: this interface defines what a help item is made of, i.e., that it has a title, a category, paragraphs, ...

File: src/main/java/org/terasology/inGameHelp/components/HelpItem.java [hunk 7]
```diff
@@ -20,13 +20,31 @@
...
 
+    /**
+     * @return the description of this HelpItem
+     */
     Iterable<ParagraphData> getParagraphs();
 
+    /**
+     * Add help information to the document
+     *
+     * @param documentdata the document for which the information is added
```
Comment (skaldarnar) [inline src/main/java/org/terasology/inGameHelp/components/HelpItem.java]: note that the parameter is modified in-place

File: src/main/java/org/terasology/inGameHelp/systems/HelpCategory.java [hunk 8]
```diff
@@ -18,14 +18,38 @@
 import org.terasology.inGameHelp.ItemsCategoryInGameHelpRegistry;
 import org.terasology.rendering.nui.widgets.browser.data.DocumentData;
 
+/**
+ * Interface that manages how a specific help tab will function
+ */
 public interface HelpCategory {
+
+	/**
+	 * Sets the registery
```
Comment (skaldarnar) [inline src/main/java/org/terasology/inGameHelp/systems/HelpCategory.java]: typo

File: src/main/java/org/terasology/inGameHelp/systems/InGameHelpCommonSystem.java [hunk 9]
```diff
@@ -62,6 +73,9 @@ public void addCategory(OnAddNewCategoryEvent event, EntityRef entity) {
         event.getCategory().setRegistry(itemsCategoryInGameHelpRegistry);
     }
 
+    /**
+     * Initialise the system
```
Comment (skaldarnar) [inline src/main/java/org/terasology/inGameHelp/systems/InGameHelpCommonSystem.java]: should contain more info on what "initialise" means

File: src/main/java/org/terasology/inGameHelp/ui/ItemWidget.java [hunk 10]
```diff
@@ -27,16 +27,31 @@
 
 import java.util.Optional;
 
+/**
+ * Base class for elements of the InGameHelp gui
+ */
 public class ItemWidget extends ItemCell {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/inGameHelp/ui/ItemWidget.java]: needs more work overall

**Comment by J0hnCena** [comment]
I appreciate the feedback. I have added the changes that you have requested. Let me know if there is anything else I need to change.

File: src/main/java/org/terasology/inGameHelp/InGameHelpCategoryRegistry.java [hunk 11]
```diff
@@ -17,7 +17,22 @@
 
 import org.terasology.inGameHelp.systems.HelpCategory;
 
+/**
+ * Interface for managing the different help categories
+ */
 public interface InGameHelpCategoryRegistry {
+
+	/**
+     * "Retrieve the {@link org.terasology.inGameHelp.systems.HelpCategory}s previously registered to this registry."
```
Comment (rzats) [inline src/main/java/org/terasology/inGameHelp/InGameHelpCategoryRegistry.java]: Misplaced comment "border" (?) here, as well as a sentence erroneously surrounded by quotes.

File: src/main/java/org/terasology/inGameHelp/InGameHelpButton.java [hunk 12]
```diff
@@ -21,6 +21,9 @@
 import org.terasology.input.Keyboard;
 import org.terasology.input.RegisterBindButton;
 
+/**
+ * Button event for opening the InGameHelp window
```
Comment (rzats) [inline src/main/java/org/terasology/inGameHelp/InGameHelpButton.java]: Would be neat to have a dot (.) at the end here (as well as everywhere else 🙂)

File: src/main/java/org/terasology/inGameHelp/components/HasBeenHelpedComponent.java [hunk 13]
```diff
@@ -17,5 +17,8 @@
 
 import org.terasology.entitySystem.Component;
 
+/**
+ * Component for when the user has been helped
```
Comment (rzats) [inline src/main/java/org/terasology/inGameHelp/components/HasBeenHelpedComponent.java]: ...and what does this imply? 😛 This particular comment needs to be slightly more specific about what the component does.

File: src/main/java/org/terasology/inGameHelp/helpCategories/GeneralHelpCategory.java [hunk 14]
```diff
@@ -29,16 +29,32 @@
...
+     *
+     * @return the name of this category
+     */
     @Override
     public String getCategoryName() {
         return "General";
     }
 
+    /**
+     * Currently does not do anything
```
Comment (rzats) [inline src/main/java/org/terasology/inGameHelp/helpCategories/GeneralHelpCategory.java]: Empty methods don't need to be documented.

File: src/main/java/org/terasology/inGameHelp/helpCategories/GeneralHelpCategory.java [hunk 15]
```diff
@@ -59,11 +75,19 @@ public DocumentData getDocumentData() {
         return documentData;
     }
 
+    /**
+     * Currently does not do anything
```
Comment (rzats) [inline src/main/java/org/terasology/inGameHelp/helpCategories/GeneralHelpCategory.java]: Stub `@Override` methods need to be NOT documented - that way they inherit the documentation of the interface that explains what the method _should_ do.

File: src/main/java/org/terasology/inGameHelp/helpCategories/ItemsCategory.java [hunk 16]
```diff
@@ -38,40 +38,65 @@
...
 import java.util.Map;
 
+/**
+ * Help category that manages the Items tab
+ */
 public class ItemsCategory implements HelpCategory {
+    
+    /** Name of this category */
     private final String name = "Items";
     private final ItemsCategoryInGameHelpRegistry itemsCategoryInGameHelpRegistry;
```
Comment (rzats) [inline src/main/java/org/terasology/inGameHelp/helpCategories/ItemsCategory.java]: The fields below this one have no documentation. TODO: add some!

File: src/main/java/org/terasology/inGameHelp/systems/InGameHelpClientSystem.java [hunk 17]
```diff
@@ -32,18 +32,30 @@
...
+ * Class that handles button events and displays the help screen to the client  
+ */
 @RegisterSystem(RegisterMode.CLIENT)
 @Share(InGameHelpClient.class)
 public class InGameHelpClientSystem extends BaseComponentSystem implements InGameHelpClient {
     @In
     NUIManager nuiManager;
 
+	/**
+	 * Initialise the system
```
Comment (rzats) [inline src/main/java/org/terasology/inGameHelp/systems/InGameHelpClientSystem.java]: Off by a few spaces 😉

File: src/main/java/org/terasology/inGameHelp/systems/InGameHelpClientSystem.java [hunk 18]
```diff
@@ -32,18 +32,30 @@
...
+	 */
     @Override
     public void initialise() {
         super.initialise();
         nuiManager.getHUD().addHUDElement("InGameHelp:UnHelpedNagWidget");
     }
 
+	/**
+	 * Handles the button event and displays the InGameHelpScreen when the button has been clicked
+     *
```
Comment (rzats) [inline src/main/java/org/terasology/inGameHelp/systems/InGameHelpClientSystem.java]: Also off by a few spaces, plus this star is misplaced.

File: src/main/java/org/terasology/inGameHelp/systems/InGameHelpClientSystem.java [hunk 19]
```diff
@@ -58,6 +70,11 @@ public void onInGameHelpButton(InGameHelpButton event, EntityRef entity) {
         }
     }
 
+	/**
+	 * Show the help screen for the document
```
Comment (rzats) [inline src/main/java/org/terasology/inGameHelp/systems/InGameHelpClientSystem.java]: And off by a few spaces again! Autoformat in IDEA to fix indentation errors like these.

**Comment by J0hnCena** [comment]
Thanks for the input. I have made the changes you have requested.

**Pull request merged**